servers/dsub/jobs/controllers/jobs_controller.py

```diff
@@ -119,55 +120,137 @@ def query_jobs(body):
+    more_jobs = _query_dstat_jobs(
+        provider, query, create_time_max, max_tasks=max_tasks)
+    # We shouldn't get more jobs than requested, but if we do ignore
+    # all the extras.
+    if len(more_jobs) > max_tasks:
+        more_jobs = more_jobs[:max_tasks]
+
+    if len(more_jobs) == 1 and max_tasks == 1:
+        # We requested one job to see if there is another page, there is
+        # so return the existing jobs and a page_token
+        return _get_query_jobs_response(jobs, query.parent_id, create_time_max)
+    elif len(more_jobs) >= max_tasks:
+        last_create_time = more_jobs[-1]['create-time']
+        penult_create_time = more_jobs[-2]['create-time']
+        if last_create_time == penult_create_time:
+            # The last job in this page has the same create-time as the first
+            # job in the next page. We need to get all jobs with this
+            # create-time and provide an offset_id with the page_token.
+            create_time_jobs = _query_dstat_jobs_single_create_time(
+                provider, query, last_create_time)
+            # Add all jobs with a different create-time to the already fetched
+            # jobs
+            create_time_cutoff = _get_create_time_index(
+                last_create_time, more_jobs)
+            jobs = jobs + more_jobs[:create_time_cutoff]
+            # Add enough jobs from the sorted list to fill the page and return
+            # a token with this create_time and the offset_id of the first job
+            # on the next page
+            needed_jobs = query.page_size - len(jobs)
+            jobs = jobs + create_time_jobs[:needed_jobs]
+            return _get_query_jobs_response(jobs, query.parent_id,
+                                            last_create_time,
+                                            create_time_jobs[needed_jobs])
+
+        else:
+            return _get_query_jobs_response(jobs + more_jobs[:-1],
+                                            query.parent_id, last_create_time)
+    else:
+        # Not enough jobs to fill the page, return everything we have
+        # and no page_token
+        return _get_query_jobs_response(jobs + more_jobs, query.parent_id)
+
+
+def _get_query_jobs_response(jobs,
+                             project_id,
+                             create_time_max=None,
+                             job_offset=None):
+    results = [_query_result(j, project_id) for j in jobs]
+    offset_id = job_offset['job-id'] + job_offset.get(
+        'task-id', '') if job_offset else None
+    token = page_tokens.encode_create_time_max(create_time_max, offset_id)
+    return QueryJobsResponse(results=results, next_page_token=token)
+
+    # If we're trying to paginate earlier than the query's minimum start time,
+    # don't return anything
+    if query.start and create_time_max and query.start > create_time_max:
+        return []
+
+
+def _query_dstat_jobs(provider, query, create_time_max, max_tasks=0):
```

Comment:
OK, so here is how I originally was thinking this **could** work in my mind, I'm thinking this could simplify the above logic (though perhaps would sacrifice some performance):

API layer: Requests pageSize+2 from generator
Generator layer (this helper method): generator function which produces an ordered stream of dsub jobs (by created, job-id+taskid)

```
def _generate_dstat_jobs():
  jobs = dstat.producer()
  last_create_time = None
  job_buffer = []
  for j in jobs:
    job_buffer.append(j)
    if j['create-time'] != last_create_time:
      for b in sorted(job_buffer, key=...): yield b
      job_buffer = []
    last_create_time = j['create-time']
  for b in sorted(job_buffer, key=...): yield b
```

Possibly this layer could also filter out according to create_time_max/offset_id. A downside is that you'd be getting rid of `max_tasks`, which might result in excessive overscanning.

The upside is that the API layer should be drastically simplified, in that it needn't worry about sorting or doing a two-phase request.

Thoughts?

Reply:
Great suggestion! It required adding one additional method to `dstat` to expose the underlying job generator but very much worth it to greatly simplify the logic in the shim. This looks much cleaner and readable now, and likely much less error-prone.

Let me know what you think.